### ISSUE TITLE
Can the generic Activator.CreateInstance<T> overload be optimized?

### ISSUE THREAD

**Issue by jamesqo** [dotnet/coreclr#7592]
Maybe instead of doing the same thing as the Type-based overload, we could simply forward to a generic method in `JitHelpers` whose IL gets substituted with initobj/newobj like [here](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Runtime/CompilerServices/jithelpers.cs#L142). Would seem to be worth it since this is an intrinsic method emitted by the C# compiler.

**Comment by jkotas** [comment]
The hard part of `Activator.CreateInstance<T>` is to make it call the right constructor. I do not think you can do it well without changing the JIT.

https://github.com/dotnet/corert/issues/368 is related.

**Comment by andrewjsaid** [comment]
This got me thinking about optimising Activator.CreateInstance<T> for generic classes with the "new()" type constraint.

It would be possible to optimise the usage of Activator.CreateInstance<T> when calling the paramtereless constructor using dynamic code compilation and static fields in generic classes (similar to how Enumerable<T>.Empty works).

I've hackup up a prototype along and a benchmark and it seems that that would produce a speed-up by a factor of 2 (and ~3 for structs) on my machine, although when talking about nanoseconds it's hardly much.

I envisage that the activator class would hold a version of my `CachedGenericConstructor<T>` class (below) as a private nested class and use that for parameterless construction.

IMPORTANT NOTE: I am using .NET 4.6.1 which might be different to .NET Core. I don't have access to .NET Core at this moment.

// * Summary *

Host Process Environment Information:
BenchmarkDotNet=v0.9.8.0
OS=Microsoft Windows NT 6.2.9200.0
Processor=Intel(R) Core(TM) i7-4710HQ CPU 2.50GHz, ProcessorCount=8
Frequency=2439993 ticks, Resolution=409.8372 ns, Timer=TSC
CLR=MS.NET 4.0.30319.42000, Arch=64-bit RELEASE [RyuJIT]
GC=Concurrent Server
JitModules=clrjit-v4.6.1586.0

Type=GenericTypeConstructor  Mode=Throughput  GarbageCollection=Concurrent Workstation

                                            Method |     Median |    StdDev |
-------------------------------------------------- |----------- |---------- |
            UsingExisting_DefaultConstructor_Class | 83.6808 ns | 1.8544 ns |
           UsingExisting_DefaultConstructor_Struct | 76.5256 ns | 1.8457 ns |
             UsingExisting_CustomConstructor_Class | 86.0214 ns | 2.9979 ns |
  UsingDynamicCompilation_DefaultConstructor_Class | 43.1595 ns | 1.6310 ns |
   UsingDynamicCompilation_CustomConstructor_Class | 44.4120 ns | 0.9563 ns |
 UsingDynamicCompilation_DefaultConstructor_Struct | 27.8055 ns | 0.2270 ns |

// ***** BenchmarkRunner: End *****

Method names:
- UsingExisting means "Activator.CreateInstance"
- UsingDynamicCompilation means the proposed code

- DefaultConstructor_Class is a class with a default constructor.
- CustomConstructor_Class is a class with a non-default parameterless constructor.
- DefaultConstructor_Struct is a struct.

The main logic is as follows


	static class CachedGenericConstructor<T> where T : new()
	{
		private static readonly Func<T> _factory;
	
		static CachedGenericConstructor()
		{
			var type = typeof (T);
			var constructorInfo = type.GetConstructor(new Type[0]);
			if (constructorInfo != null)
			{
				Expression newExpression = Expression.New(constructorInfo);
				var lambdaExpression = Expression.Lambda<Func<T>>(newExpression);
				var function = lambdaExpression.Compile();
				_factory = function;
			}
			else if (type.IsValueType)
			{
				Expression defaultExpression = Expression.Default(type);
				var lambdaExpression = Expression.Lambda<Func<T>>(defaultExpression);
				var function = lambdaExpression.Compile();
				_factory = function;
			}
			else
			{
				throw new InvalidOperationException("Not sure that this is possible...");
			}
		}
	
		public static T Construct() => _factory();
	
	}

**Comment by jkotas** [comment]
We would not want to (and it is not even possible today) to take dependency on Linq expressions from CoreLib. 

This optimization should be done in AOT-friendly way - without dynamic code generation.

**Comment by karelz** [comment]
I'll be brave and say that this is not worth the effort. And close it. Feel free to push back if you think otherwise.

**Comment by karelz** [comment]
I let @jkotas to comment if you changed his mind. This is way out of my area of expertise ...
cc @adamsitnik @steveharter @GrabYourPitchforks

**Comment by GrabYourPitchforks** [comment]
I had hoped that https://github.com/dotnet/corefx/issues/24390 would be a way to address this. Essentially, you could ask the runtime for a delegate which acts as a fact factory for some type _T_, then you could cache that delegate and keep calling it as needed.

**Comment by jkotas** [comment]
It would be certainly better to do this without runtime codegen (i.e. without using Reflection.Emit or equivalent) using the technique used by .NET Native. This requires change in the JIT that only a very few folks are able to do.

We have given up to large degree on making CoreCLR to be low footprint and AOT friendly runtime, so dependency on `Reflection.Emit` from `Activator.CreateInstance` would not prohibitive. Anybody interested in submitting a PR? The implementation should basically emit the delegate that @GrabYourPitchforks described and cache in a static.

**Comment by GrabYourPitchforks** [comment]
TBH I don't think we'd need ref emit. Eventually a reflection-based instantiation turns into a call to [`AllocateObject`](https://github.com/dotnet/coreclr/blob/d3e39bc2f81e3dbf9e4b96347f62b49d8700336c/src/vm/gchelpers.cpp#L1140). If we can perform all of the Type checks up front and cache the pointer that needs to be passed to this method, we could make a very fast implementation of `GetUninitializedObject`. A fast factory would then wrap a call to this followed by a `calli` call to the object ctor.

**Comment by jkotas** [comment]
Right, this is pretty close what the ultimate implementation with the help of the JIT would do. It would still need some new intrinsics and such.  If we took that route, I think it is better to do the proper AOT-friendly fix that does what you have described as part of JITing the method.

**Comment by YohDeadfall** [comment]
I wrote a scratch yesterday, but haven't tested it or even compiled. It's near the same @GrabYourPitchforks said, so I can take the issue and will open a PR later.

As @jkotas wrote in https://github.com/dotnet/coreclr/issues/6843#issuecomment-241566379, this code doesn't support default constructors for value types, but will fix it.

```csharp
public static class Activator
{
    private static readonly MethodInfo s_createInstanceBoxedMethod = new Func<int>(CreateInstanceBoxed).Method.GetGenericMethodDefinition();
    private static object CreateInstanceBoxed<T>(bool nonPublic, bool wrapExceptions) =>
        CreateInstance<T>(nonPublic, wrapExceptions);

    public static object CreateInstance(Type type) => CreateInstance(type, nonPublic);

    public static object CreateInstance(Type type, bool nonPublic)
    {
        if (type == null)
            throw new ArgumentNullException(nameof(type));

        // This methods should be cached like it's going now
        return ((Func<object>)s_createInstanceBoxedMethod.MakeGenericMethod(type).CreateDelegate(typeof(Func<object>)));
    }

    public static T CreateInstance<T>() => CreateInstance<T>(false, true);

    private static T CreateInstance<T>(bool nonPublic, bool wrapExceptions)
    {
        if (ActivatorCahce<T>.HasElementType)
            throw new MissingMethodException(SR.Arg_NoDefCTor);

        if (ActivatorCahce<T>.IsValueType)
            return default;

        if (ActivatorCache<T>.HasPublicConstructor || nonPublic)
        {
            object instance = RuntimeTypeHandle.Allocate(this);

            try
            {
                   ActivatorCache<T>.Constructor(instance);
            }
            catch (Exception exception) when (wrapExceptions)
            {
                   throw new TargetInvocationException(exception);
            }

            return instance;
        }

        throw new MissingMethodException(SR.Arg_NoDefCTor);
    }

    private static class ActivatorCache
    {
        private static readonly ConstructorInfo s_delegateCtorInfo = typeof(CtorDelegate).GetConstructor(new [] { typeof(object), typeof(IntPtr) });

        public static CtorDelegate CreateConstructor(ConstructorInfo ctorInfo)
        {
            RuntimeMethodHandle ctorHandle = ctorInfo.RuntimeMethodHandle;
            return s_delegateCtorInfo.Invoke(null, RuntimeMethodHandle.GetFunctionPointer(ctorHandle));
        }
    }

    private static class ActivatorCache<T>
    {
        public static readonly bool IsValueType;
        public static readonly bool HasElementType;
        public static readonly bool HasPublicConstructor;
        public static readonly CtorDelegate Constructor;

        static ActivatorCache()
        {
            var runtimeType = typeof(T);

            IsValueType = runtimeType.IsValueType;
            HasElementType = runtimeType.HasElementType;

            if (IsValueType || HasElementType)
            {
                   return;
            }

            ConstructorInfo runtimeCtorInfo = runtimeType.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, new [] { typeof(T) }, null);

            if (runtimeCtorInfo != null)
            {
                   HasPublicConstructor = runtimeCtorInfo.IsPublic;
                   Constructor = ActivatorCache.CreateConstructor(runtimeCtorInfo);
            }
        }
    }
}
```

**Comment by yahorsi** [comment]
>We have given up to large degree on making CoreCLR to be low footprint and AOT friendly runtime
Does it means CoreRT dead? Are there any roadmaps or plans for releasing CoreRT?

**Comment by jkotas** [comment]
The code [shared](https://github.com/dotnet/coreclr/tree/master/src/System.Private.CoreLib/shared) between runtimes should stay to be AOT-friendly. My note applies to non-shared code that is specific to CoreCLR.

> Are there any roadmaps or plans for releasing CoreRT?

https://github.com/dotnet/corert/issues/7200

**Comment by yahorsi** [comment]
Thank you

**Comment by YohDeadfall** [comment]
Resolved via #25145 by @jkotas.

**Comment by karelz** [comment]
Fixed by #25145
